libvast/builtins/connectors/file.cpp

```diff
@@ -0,0 +1,268 @@
+      } else if (arg == "-f" || arg == "--follow") {
+        following = true;
+      } else if (not arg.starts_with("-")) {
+        std::error_code err{};
+        auto status = std::filesystem::status(arg, err);
+        if (err) {
+          return caf::make_error(ec::parse_error,
+                                 fmt::format("could not parse file path {}: {}",
+                                             arg, err));
+        }
+        auto exists = (status.type() != std::filesystem::file_type::not_found);
+        if (not exists) {
+          return caf::make_error(ec::filesystem_error,
+                                 fmt::format("file {} does not exist", arg));
+        }
+        if (not path.empty()) {
+          return caf::make_error(ec::syntax_error,
+                                 fmt::format(
+                                   "multiple paths have been "
+                                   "specified; currently only one "
+                                   "path per read operator call is valid"));
+        }
+        auto permissions = status.permissions();
+        auto readable = ((permissions & std::filesystem::perms::owner_read)
+                           != std::filesystem::perms::none
+                         && (permissions & std::filesystem::perms::group_read)
+                              != std::filesystem::perms::none
+                         && (permissions & std::filesystem::perms::others_read)
+                              != std::filesystem::perms::none);
+        if (not readable) {
+          return caf::make_error(ec::filesystem_error,
+                                 fmt::format("file {} is not readable", arg));
+        }
+        path = arg;
+      } else {
+        return caf::make_error(
+          ec::invalid_argument,
+          fmt::format("unexpected argument for 'file' connector: {}", arg));
+      }
+    }
+    if (path.empty()) {
+      return caf::make_error(ec::syntax_error,
+                             fmt::format("no file specified"));
+    }
+    auto fd = STDIN_FILENO;
+    switch (file_type) {
+      default:
+        return caf::make_error(ec::filesystem_error, "unsupported file type",
+                               path);
+      case std::filesystem::file_type::socket: {
+        if (path == stdin_path) {
+          return caf::make_error(ec::filesystem_error,
+                                 "cannot use STDIN as UNIX "
+                                 "domain socket");
+        }
+        auto uds = detail::unix_domain_socket::connect(path);
+        if (!uds) {
+          return caf::make_error(ec::filesystem_error,
+                                 "failed to connect to UNIX domain socket at",
+                                 path);
+        }
+        fd = uds.recv_fd(); // Blocks!
+        break;
+      }
+      case std::filesystem::file_type::regular: {
+        if (path != stdin_path) {
+          fd = ::open(path.c_str(), std::ios_base::binary | std::ios_base::in);
```

Comment:
This is pretty much TOCTOU. Can we just try to open it without doing the access checks?

Reply:
All checks have been removed (except for `is_socket`) in https://github.com/tenzir/vast/pull/3088/commits/a3b046c402783025a72fa735ac583f17f6598ba3